### ISSUE TITLE
Support random values

### ISSUE THREAD

**Issue by nicolaiparlog** [junit-pioneer/junit-pioneer#49]
Support use cases for random values, e.g. as parameters (like [this](https://github.com/CodeFX-org/demo-junit-5/blob/master/src/main/java/org/codefx/demo/junit5/RandomProvider.java); inspired by junit-team/junit5#839) or fields or arguments for parameterized tests.

**Comment by sormuras** [comment]
And/or as a simple `ParameterResolver` [implementation](
https://github.com/junit-team/junit5-samples/blob/master/junit5-jupiter-extensions/src/main/java/com/example/random/RandomParametersExtension.java) show-cased by the _junit5-sample_:

Usage could look like:
```java
@ExtendWith(RandomParametersExtension.class)
class RandomParametersExtensionTests {
  @Test
  void injectsInteger(@Random int i, @Random int j) { assertNotEquals(i, j); }
  @Test
  void injectsDouble(@Random double d) { assertEquals(0.0, d, 1.0); }
}
```

**Comment by nicolaiparlog** [comment]
My personal opinion is that however this is implemented, it should be possible to repeat the tests with the same random values. That makes it necessary:

* to have a deterministic entry point that spits out the seed value that is used
* to allow configuration of that entry point to pass the seed to it

As far as I see it (and as long as JUnit does not randomize test order), the only challenge for the deterministic entry point is multi-threading. I am willing to accept that limitation.

Speaking of randomizing test order, it would be interesting to find out whether an extension could do that.

**Comment by sormuras** [comment]
> [...] repeat the tests with the same random values.

`@Random(seed = 123)` should suffice. When running with the default `@Random` it could be reported via the [TestReporter](https://junit.org/junit5/docs/current/api/org/junit/jupiter/api/TestReporter.html) facility.

> Speaking of randomizing test order, it would be interesting to find out whether an extension could do that.

https://github.com/junit-team/junit5/issues/13 is still open...

**Comment by nicolaiparlog** [comment]
Possible features:

* set seed per class
* set seed globally via configuration parameter

**Comment by Michael1993** [comment]
I don't quite see the use-case for this. You should always control your input to your tests which means no randomizing.

**Comment by Bukama** [comment]
> I don't see the use-case for this. You should always control your input to your tests which means no randomizing.

I'm with @Michael1993.

* If I don't need a value for my method I can pass whatever class type of the parameter / class under test  allows (null, default, any number, etc.)
* If I want to check for edge cases code coverage tools like [JaCoCo](https://www.jacoco.org/) or mutation testing tools like [PIT](https://pitest.org/) take this spot
* If I really want to use random values, we don't have to reinvent the wheel, as there are plenty of property based testing tools like [jqwik](https://jqwik.net/property-based-testing.html)

So I vote to close this issue.

**Comment by beatngu13** [comment]
> I don't see the use-case for this. You should always control your input to your tests which means no randomizing.

Actually, random testing is quite a big thing. Be it at the top of the test automation pyramid (e.g. [Sapienz](https://engineering.fb.com/developer-tools/sapienz-intelligent-automated-software-testing-at-scale/)) or at the bottom (e.g. [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck)).

However, let's not get too philosophical. 😉 I agree with @Bukama that we shouldn't do this, there are excellent libs already available. I think this is out of scope and would open a can of worms. Therefore, I also vote to close this issue.

**Comment by nicolaiparlog** [comment]
Whoa, why does everybody want to close this? No random values in tests? Where's your sense of adventure? :rocket:

I'd still like to implement this eventually:

* there is an actual JUnit user asking for this feature
* with a seed, the randomness is reproducible, so there's no disadvantage
* switching to a different test engine or even entire testing tool has a slightly bigger foot print than adding an extension
* it sounds like fun!

Vote not to close :stuck_out_tongue_winking_eye:

**Comment by nicolaiparlog** [comment]
> I say never be complete, I say stop being perfect, I say let... lets evolve, let the chips fall where they may.

> Introduce a little anarchy. Upset the established order and everything becomes chaos. I'm an agent of chaos. Oh, and you know the thing about chaos?  It's fair.

> Why so serious? 🃏

Yes, quoting Tyler Durden and the Joker is juvenile and both are full of shit, but... it's fun. :)

**Comment by nipafx** [comment]
I also just now realize that it may be interesting to think about this in terms of parameterized tests:

* this could be a source for parameterized test arguments
* this could be a test template of its own, spawning a preconfigured number of tests with random parameters

I'm gonna unilaterally green-ligh this. 😁

**Comment by nipafx** [comment]
To be a good issue for first contributors, this issue should describe the specific annotations to be created as well as give a rough idea how to implement them.

**Comment by keturn** [comment]
When you get to the point of wanting to have reproducible runs, this looks like it might be of use: [java.util.SplittableRandom](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/SplittableRandom.html)

**Comment by nipafx** [comment]
In lieu of fleshing out this issue, I created [a proof of concept](https://github.com/junit-pioneer/junit-pioneer/tree/lab/random-arguments). 😁

**Comment by nipafx** [comment]
I think the best use for this extension is to become part of the revamped Cartesian product extension, so I don't think it makes sense to invest more time here before #415 is done.

**Comment by tubbynl** [comment]
In randomizing data JavaFaker (https://github.com/DiUS/java-faker) does provide sensible random data, in combination with parameterized tests you do get valuable edge-case information. 

Certainly when testing validation stuff.

**Comment by etrandafir93** [comment]
hello @nipafx this issue still relevant? 

I know it has been discussed a long time ago, but IMO this doesn't make much sense **without** a parameterizable seed because we'd lose the reproducibility.
 
So I think the way jqwik does it is correct: you can run a test with/without providing a seed. If it fails, the seed that was used will **always** logged, allowing you to rerun the test and confirm the issue was fixed

**Comment by Bukama** [comment]
As a note, @Michael1993 opened an experimenting draft about this issue a few days ago, see #831.

**Comment by etrandafir93** [comment]
I see, thanks for the quick reply @Bukama! I had in mind something smaller and simpler.. just a custom annotation that is linked to a `ArgumentsProvider` and can be used at a method level without any reflection.  

It would be much simpler (just 1 annotation + 1 class for the arg provider) but not as flexible. If this sounds interesting, I don't mind pushing it to a branch :)
